### ISSUE TITLE
Concept Doc Guidelines

### ISSUE THREAD

**Issue by tajmone** [tajmone/PBCodeArcProto#12]
Thanks for the `Concept.md` doc, this is an important reference document. So much so that I've copied it over to this repo:

- https://github.com/tajmone/PBCodeArcProto/blob/master/Concept.md

The idea is to use this Issue to discuss its points, and use the copy in this repo to actually edit the document — unlike your upstream repository, this is a temporary repository that will be eventually deleted, so we have more freedom in editing the document, and even free to include comments and questions directly in the doc.

I'd like to go over some of the points of the document in this thread.

**Comment by tajmone** [comment]
## PBI Files

> - `.pbi` (PureBasic include file) should only be used for code that doesn't contain code that is executed directly. All commands should be encapsulated in procedures, like libraries (`.dll` (Windows), `.so` (Linux) and `.dylib` (Mac)).

Isn't this limiting? Sometimes we might want to include a file that handles cross-platform settings, intended to be executed at the place of inclusion (eg, definined constants, vars, macros and other helpers for cross platfromerness).

Also, when it comes to modules, the above definition becomes ambiguos.

Maybe a variation draft of the above distinctions:

> - What is the difference between "`*.pb`" and "`*.pbi`" files?
> 
>     - `.pb` (PureBasic main source file) — A compilable sourcecode file (either standalone or depending on other include files) which can' t be used as an include file.
>     
>     - `.pbi` (PureBasic include file) — A source file intended to be included (directly or indirectly) as an external resource by a main `.pb` source file which is the entry point of compilation.
>      
>         Include files are not intended to be executed as standalone code. However, they may contain a usage example or test code within a `CompilerIf #PB_Compiler_MainFile` block, which will only be executed if the `.pbi` file is executed directly (ie, not included) and whose purpose is to provide a demonstration on how to use the functionality offered by the include file.
> 
>         > __NOTE__: All the code inside the `CompilerIf #PB_Compiler_MainFile` will be completely ignored by the compiler when the `.pbi` file is being included by another source file, therefore it will not affect the binary compiled file.
>      
>         Since they are intended to be shared resources,  `.pbi` files usually encapsulated  their functionality in procedures — like libraries do (`.dll` (Windows), `.so` (Linux) and `.dylib` (Mac)) — or even modules, allow full control to the inclduing file as to when to invoke its functionality.

The difference here is that it clarifies the context and purpose of the two file extensions, so it's _descriptive_ rather then _prescriptive_, thus allowing some margin of freedom for special cases which might not fit the description (as long as they fall within the scope and don't conflict, it should be fine).

**Comment by SicroAtGit** [comment]
> Isn't this limiting? Sometimes we might want to include a file that handles cross-platform settings, intended to be executed at the place of inclusion (eg, definined constants, vars, macros and other helpers for cross platfromerness).

> Also, when it comes to modules, the above definition becomes ambiguos.

Yes, I haven't found the right formulation yet. Of course, what you mention should be allowed.

I simply want that `pbi` files can be inserted blindly and carefree into programs. If a user later reports to the developer of the program that the program fills his PC with strange files or connects to questionable Internet servers, this would be very embarrassing for the developer.
